src/hooks/useApproveCallback.ts

```diff
@@ -103,23 +135,85 @@ export function useApproveCallback(
 
 // wraps useApproveCallback in the context of a swap
 export function useApproveCallbackFromTrade(
-  trade: V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType> | undefined,
+  trade:
+    | V2Trade<Currency, Currency, TradeType>
+    | V3Trade<Currency, Currency, TradeType>
+    | Trade<Currency, Currency, TradeType>
+    | undefined,
   allowedSlippage: Percent
 ) {
   const { chainId } = useActiveWeb3React()
-  const v3SwapRouterAddress = chainId ? SWAP_ROUTER_ADDRESSES[chainId] : undefined
   const amountToApprove = useMemo(
     () => (trade && trade.inputAmount.currency.isToken ? trade.maximumAmountIn(allowedSlippage) : undefined),
     [trade, allowedSlippage]
   )
-  return useApproveCallback(
+
+  const approveCallback = useApproveCallback(
     amountToApprove,
     chainId
       ? trade instanceof V2Trade
         ? V2_ROUTER_ADDRESS[chainId]
         : trade instanceof V3Trade
-        ? v3SwapRouterAddress
-        : undefined
+        ? V3_ROUTER_ADDRESS[chainId]
```

Comment:
trade will only be a `V3Trade` if its from legeacy router logic?

Reply:
hm in this case actually we pass a v2 | v3 | trade which is derived from the routing api `Trade` by looking at approvals. So if a user has already approved the v3 router, we convert the `Trade` into a `V3Trade`